gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFSInputStream.java

```diff
@@ -75,31 +80,70 @@ class GoogleHadoopFSInputStream extends FSInputStream implements IOStatisticsSou
 
   // Statistic tracker of the Input stream
   private final GhfsInputStreamStatistics streamStatistics;
+  private final VectoredIOImpl vectoredIO;
 
   static GoogleHadoopFSInputStream create(
-      GoogleHadoopFileSystem ghfs, URI gcsPath, FileSystem.Statistics statistics)
+      GoogleHadoopFileSystem ghfs,
+      URI gcsPath,
+      VectoredReadOptions vectoredReadOptions,
+      FileSystem.Statistics statistics)
       throws IOException {
     logger.atFiner().log("create(gcsPath: %s)", gcsPath);
     GoogleCloudStorageFileSystem gcsFs = ghfs.getGcsFs();
-    SeekableByteChannel channel =
-        gcsFs.open(gcsPath, gcsFs.getOptions().getCloudStorageOptions().getReadChannelOptions());
-    return new GoogleHadoopFSInputStream(ghfs, gcsPath, channel, statistics);
+    FileInfo fileInfo = null;
+    SeekableByteChannel channel;
+    // Extract out the fileInfo call here and use it in readChannel as well as in vectoredRead API
+    if (shouldPreFetchFileInfo(gcsFs.getOptions())) {
+      // ingest the fileInfo extracted while creating gcsio channel to avoid duplicate call.
+      fileInfo = gcsFs.getFileInfoObject(gcsPath);
+      channel =
+          gcsFs.open(fileInfo, gcsFs.getOptions().getCloudStorageOptions().getReadChannelOptions());
+    } else {
+      // cases where fileInfo wouldn't have been requested in gcsio layer.
+      channel =
+          gcsFs.open(gcsPath, gcsFs.getOptions().getCloudStorageOptions().getReadChannelOptions());
+    }
+    return new GoogleHadoopFSInputStream(
+        ghfs, gcsPath, fileInfo, channel, vectoredReadOptions, statistics);
+  }
+
+  private static boolean shouldPreFetchFileInfo(GoogleCloudStorageFileSystemOptions gcsFSOptions) {
+    // FileInfo is requested while opening the channel in gcsio channel layer in following
+    // conditions
+    // 1. failFastOnNotFound is enabled
+    // 2. java-storage library is in use (failedFast in no-op for grpc flow).
+    // prefecthing the fileInfo in FsInputSteam. So, that it can be used across other read API i.e.
+    // vectoredRead
+    if (gcsFSOptions.getClientType() == ClientType.STORAGE_CLIENT
+        || gcsFSOptions
+            .getCloudStorageOptions()
+            .getReadChannelOptions()
+            .isFastFailOnNotFoundEnabled()) {
+      return true;
+    }
+    return false;
   }
 
   static GoogleHadoopFSInputStream create(
-      GoogleHadoopFileSystem ghfs, FileInfo fileInfo, FileSystem.Statistics statistics)
+      GoogleHadoopFileSystem ghfs,
+      FileInfo fileInfo,
+      VectoredReadOptions vectoredReadOptions,
+      FileSystem.Statistics statistics)
       throws IOException {
     logger.atFiner().log("create(fileInfo: %s)", fileInfo);
     GoogleCloudStorageFileSystem gcsFs = ghfs.getGcsFs();
     SeekableByteChannel channel =
```

Comment:
Vectored IO is not a new feature which connector can enable or disable. It's API is already there as part of hadoop 3.3.5. So, technically we support `readVectored` API. Via this implementation we are optimizing it.

We will still be needign the `channel` to serve regular `read` calls. Customer can use `read` and `readVectored` API on same InputStream. Only contract which will be broken is the `position` value can't be promised to have correct value if readVectored is called after `read`. 
https://hadoop.apache.org/docs/current/api/org/apache/hadoop/fs/PositionedReadable.html#readVectored-java.util.List-java.util.function.IntFunction-:~:text=The%20position%20returned%20by%20getPos()%20after%20readVectored()%20is%20undefined.

Reply:
As the code flow is totally independent I don't see a much need for it.

Reply:
I think this can help us detect bugs. If this scenario come and they get corrupted data, it will be hard to troubleshoot, I think. I will leave it up to you.